File: 329-longest-increasing-path-in-a-matrix/longest-increasing-path-in-a-matrix.cpp
class Solution {
public:
    int solve(vector<vector<int>> &matrix, int i, int j, int par, vector<vector<int>> &dp){
        int m = matrix.size();
        int n = matrix[0].size();
        if(i < 0 || i >= m || j < 0 || j >= n) return 0;
        if(matrix[i][j] <= par) return 0;
        if(dp[i][j] != -1) return dp[i][j];
        // vis[i][j] = 1;
        dp[i][j] = 1 + max({solve(matrix, i, j+1, matrix[i][j], dp), solve(matrix, i, j-1, matrix[i][j], dp), solve(matrix, i+1, j, matrix[i][j], dp), solve(matrix, i-1, j, matrix[i][j], dp)});
        // vis[i][j] = 0;
        return dp[i][j];
    }
    int longestIncreasingPath(vector<vector<int>>& matrix) {
        int m = matrix.size();
        int n = matrix[0].size();
        int maxi = INT_MIN;
        vector<vector<int>> dp(m, vector<int>(n, -1));
        for(int i=0; i<m; i++){
            for(int j=0; j<n; j++){
                // if(i == 1 && j == n-1) cout<<dp[i][j]<<endl;
                // vector<vector<int>> vis(m, vector<int>(n, 0));
                maxi = max(maxi, solve(matrix, i, j, -1, dp));
                // if(i == 1 && j == n-1) cout<<dp[i][j]<<endl;
            }
        }
        return maxi;
    }
};

File: 329-longest-increasing-path-in-a-matrix/longest-increasing-path-in-a-matrix.py
class Solution:
    def longestIncreasingPath(self, matrix: List[List[int]]) -> int:

        m, n = len(matrix), len(matrix[0])
        dp = [[-1] * (n+1) for _ in range(m+1)]

        visited = [[0] * (n+1) for _ in range(m+1)]

        def dfs(i ,j):
            if i < 0 or i >= m or j < 0 or j >= n:
                return 0
            if visited[i][j] == 1:
                return 0
            if dp[i][j] != -1:
                return dp[i][j]
            
            offset = [(0, 1), (0, -1), (1,0), (-1,0)]
            path = 1
            visited[i][j] = 1
            for x,y in offset:
                if 0 <= (i+x) < m and 0 <= (j+y) < n and visited[i+x][j+y] == 0 and matrix[i+x][j+y] > matrix[i][j]:
                    path = max(path, 1 + dfs(i+x, j+y))
            visited[i][j] = 0
            
            dp[i][j] = path
            return dp[i][j]

        longest_path = 0
        
        for i in range(m):
            for j in range(n):
                longest_path = max(longest_path, dfs(i, j))
        
        return longest_path
                
                
        